app/backends/workflow-orchestration.ts

```diff
@@ -0,0 +1,227 @@
+import { NextFunction, Response } from 'express';
+import { Logger } from 'winston';
+import db, { Transaction } from '../util/db';
+import env from '../util/env';
+import { readCatalogItems } from '../util/stac';
+import HarmonyRequest from '../models/harmony-request';
+import { Job, JobStatus } from '../models/job';
+import JobLink from '../models/job-link';
+import WorkItem, { getNextWorkItem, WorkItemStatus, updateWorkItemStatus, getWorkItemById, workItemCountForStep } from '../models/work-item';
+import { getWorkflowStepByJobIdStepIndex } from '../models/workflow-steps';
+import { rmdir } from '../util/file';
+
+const MAX_TRY_COUNT = 1;
+const RETRY_DELAY = 1000;
+
+/**
+ * Return a work item for the given service
+ * @param req - The request sent by the client
+ * @param res - The response to send to the client
+ * @param next - The next function in the call chain
+ * @returns Resolves when the request is complete
+ */
+export async function getWork(
+  req: HarmonyRequest, res: Response, next: NextFunction, tryCount = 1,
+): Promise<void> {
+  const { serviceID } = req.query;
+  const { logger } = req.context;
+  if (tryCount === 1) {
+    logger.debug(`Getting work for service [${serviceID}]`);
+  }
+  let workItem;
+  await db.transaction(async (tx) => {
+    workItem = await getNextWorkItem(tx, serviceID as string);
+  });
+  if (workItem) {
+    res.send(workItem);
+  } else if (tryCount < MAX_TRY_COUNT) {
+    setTimeout(async () => {
+      getWork(req, res, next, tryCount + 1);
```

Comment:
There's this type of call to setTimeout here and in the service wrapper code. I don't fully understand it, but I think James does and the logic does seem to work. The issue we were seeing is that when we used long polling the k8s stack networking failure rates were high. I'd say somewhere around 25% of network calls from our services to the harmony 3001 port would fail to resolve. This posts talks about a race condition with opening and reserving a port when multiple requests on the same k8s node attempt to initiate a connection: https://tech.xing.com/a-reason-for-unexplained-connection-timeouts-on-kubernetes-docker-abd041cf7e02.

So are you suggesting that we might just need to change this to `await getWork(req...)`?

Reply:
Bummer that it runs deeper.  Yes to `await getWork(req...)`, though

Reply:
I changed it - tried to get long polling working, but I'm seeing most of my synchronous requests failing indicating we still have this k8s/docker desktop networking issue. So we'll need a separate ticket to get long polling working reliably.